lib/AST/AccessRequests.cpp

```diff
@@ -0,0 +1,349 @@
+//===--- AccessRequests.cpp - AccessLevel and AccessScope Requests --------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Subsystems.h"
+#include "swift/AST/AccessScopeChecker.h"
+#include "swift/AST/AccessRequests.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/DiagnosticsCommon.h"
+#include "swift/AST/LazyResolver.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/Types.h"
+
+#include "llvm/Support/MathExtras.h"
+
+using namespace swift;
+
+namespace swift {
+// Implement the access-control type zone.
+#define SWIFT_TYPEID_ZONE SWIFT_ACCESSS_REQUESTS_TYPEID_ZONE
+#define SWIFT_TYPEID_HEADER "swift/AST/AccessTypeIDZone.def"
+#include "swift/Basic/ImplementTypeIDZone.h"
+#undef SWIFT_TYPEID_ZONE
+#undef SWIFT_TYPEID_HEADER
+}
+
+//----------------------------------------------------------------------------//
+// AccessLevel computation
+//----------------------------------------------------------------------------//
+AccessLevel AccessLevelRequest::evaluate(Evaluator &evaluator,
+                                         ValueDecl *D) const {
+  assert(!D->hasAccess());
+
+  // Check if the decl has an explicit access control attribute.
+  if (auto *AA = D->getAttrs().getAttribute<AccessControlAttr>())
+    return AA->getAccess();
+
+  // Special case for accessors, which inherit the access of their storage.
+  // decl. A setter attribute can also override this.
+  if (auto accessor = dyn_cast<AccessorDecl>(D)) {
+    AbstractStorageDecl *storage = accessor->getStorage();
+    switch (accessor->getAccessorKind()) {
+    case AccessorKind::Get:
+    case AccessorKind::Address:
+      return storage->getFormalAccess();
+    case AccessorKind::Set:
+    case AccessorKind::MutableAddress:
+    case AccessorKind::MaterializeForSet:
+      return storage->getSetterFormalAccess();
+    case AccessorKind::WillSet:
+    case AccessorKind::DidSet:
+      // These are only needed to synthesize the setter.
+      return AccessLevel::Private;
+    }
+  }
+
+  // Special case for associated types: inherit access from protocol.
+  if (auto assocType = dyn_cast<AssociatedTypeDecl>(D)) {
+    auto prot = assocType->getProtocol();
+    return std::max(prot->getFormalAccess(), AccessLevel::Internal);
+  }
+
+  // Special case for dtors and enum elements: inherit from container
+  if (D->getKind() == DeclKind::Destructor ||
+      D->getKind() == DeclKind::EnumElement) {
+    if (D->isInvalid()) {
```

Comment:
I know the `isInvalid()` check is pre-existing, but can we do this without that check? For example, do a `dyn_cast<NominalTypeDecl>` below, and if it's null, return `AccessLevel::Private`.

Reply:
Go ahead and land! We can clean this up later.